### ISSUE TITLE
One node is unavailable in discovery mode causes a new node cannot join the cluster

### ISSUE THREAD

**Issue by wujushan** [rqlite/rqlite#923]
I am developing by using the verison **6.7.0**. I find that, in discovery mode, if one node is unavailable will cause a new node cannot join the cluster. because the new node uses the IP of that unavailable node. 

I deployed the discovery service [rqlite-disco-express](https://github.com/rqlite/rqlite-disco-express) in my own PC. 

The log below is printed by new node. I find that the new node always uses the incorrect ip(**172.16.160.103:4001**) to join the cluster rather than trys the other one. 

**Please try a few more times if you cannot get the error log like the one below.**

```
rqlited] 2021/11/09 08:20:17 rqlited starting, version 6, commit unknown, branch unknown, compiler gc
[rqlited] 2021/11/09 08:20:17 go1.17.2, target architecture is arm, operating system target is linux
[rqlited] 2021/11/09 08:20:17 launch command: /bin/rqlited -http-addr 0.0.0.0:4001 -raft-addr 0.0.0.0:4002 -http-adv-addr 172.16.160.49:4001 -raft-adv-addr 172.16.160.49:4002 -disco-id e2d437c0-3d66-11ec-8442-8fa65a76e101 --disco-url http://172.16.160.49:10808 -on-disk /rqlite/file/data
[mux] 2021/11/09 08:20:17 mux serving on [::]:4002, advertising 172.16.160.49:4002
[rqlited] 2021/11/09 08:20:17 Raft TCP mux Listener registered with 1
[rqlited] 2021/11/09 08:20:17 preexisting node state detected in /rqlite/file/data
[rqlited] 2021/11/09 08:20:17 registering with Discovery Service at http://172.16.160.49:10808 with ID e2d437c0-3d66-11ec-8442-8fa65a76e101
[discovery] 2021/11/09 08:20:17 discovery client attempting registration of 172.16.160.49:4001 at http://172.16.160.49:10808/e2d437c0-3d66-11ec-8442-8fa65a76e101
[discovery] 2021/11/09 08:20:17 discovery client successfully registered 172.16.160.49:4001 at http://172.16.160.49:10808/e2d437c0-3d66-11ec-8442-8fa65a76e101
[rqlited] 2021/11/09 08:20:17 Discovery Service responded with nodes: [172.16.160.43:4001 172.16.160.49:4001 172.16.160.103:4001]
[rqlited] 2021/11/09 08:20:17 join addresses specified, node is not bootstrapping
[store] 2021/11/09 08:20:17 opening store with node ID 172.16.160.49:4002
[store] 2021/11/09 08:20:17 configured for an on-disk database at /rqlite/file/data/db.sqlite
[store] 2021/11/09 08:20:17 ensuring directory at /rqlite/file/data exists
[store] 2021/11/09 08:20:17 ensuring directory at /rqlite/file/data exists
[store] 2021/11/09 08:20:17 0 preexisting snapshots present
[store] 2021/11/09 08:20:17 first log index: 1, last log index: 16, last command log index: 11:
[store] 2021/11/09 08:20:17 created in-memory database at open
2021-11-09T08:20:17.153Z [INFO]  raft: initial configuration: index=4 servers="[{Suffrage:Voter ID:172.16.160.49:4002 Address:172.16.160.49:4002} {Suffrage:Voter ID:172.16.160.43:4002 Address:172.16.160.43:4002} {Suffrage:Voter ID:172.16.160.103:4002 Address:172.16.160.103:4002}]"
[store] 2021/11/09 08:20:17 no cluster bootstrap requested
[cluster] 2021/11/09 08:20:17 service listening on 172.16.160.49:4002
[rqlited] 2021/11/09 08:20:17 Cluster TCP mux Listener registered with 2
2021-11-09T08:20:17.154Z [INFO]  raft: entering follower state: follower="Node at 172.16.160.49:4002 [Follower]" leader=
[http] 2021/11/09 08:20:17 service listening on [::]:4001
[rqlited] 2021/11/09 08:20:17 join addresses are: [172.16.160.43:4001 172.16.160.103:4001]
[cluster-join] 2021/11/09 08:20:17 failed to join cluster at [172.16.160.43:4001 172.16.160.103:4001]: Post "http://172.16.160.103:4001/join": dial tcp 172.16.160.103:4001: connect: connection refused, sleeping 5s before retry
[store] 2021/11/09 08:20:17 4 committed log entries applied in 2.468035ms, took 46.519746ms since open
[store] 2021/11/09 08:20:17 successfully switched to on-disk database
2021-11-09T08:20:18.486Z [WARN]  raft: heartbeat timeout reached, starting election: last-leader=172.16.160.43:4002
2021-11-09T08:20:18.486Z [INFO]  raft: entering candidate state: node="Node at 172.16.160.49:4002 [Candidate]" term=2830
2021-11-09T08:20:18.498Z [ERROR] raft: failed to make requestVote RPC: target="{Voter 172.16.160.103:4002 172.16.160.103:4002}" error="dial tcp 172.16.160.103:4002: connect: connection refused"
2021-11-09T08:20:18.537Z [INFO]  raft: election won: tally=2
2021-11-09T08:20:18.537Z [INFO]  raft: entering leader state: leader="Node at 172.16.160.49:4002 [Leader]"
2021-11-09T08:20:18.537Z [INFO]  raft: added peer, starting replication: peer=172.16.160.43:4002
2021-11-09T08:20:18.537Z [INFO]  raft: added peer, starting replication: peer=172.16.160.103:4002
2021-11-09T08:20:18.539Z [INFO]  raft: pipelining replication: peer="{Voter 172.16.160.43:4002 172.16.160.43:4002}"
2021-11-09T08:20:18.539Z [ERROR] raft: failed to appendEntries to: peer="{Voter 172.16.160.103:4002 172.16.160.103:4002}" error="dial tcp 172.16.160.103:4002: connect: connection refused"
2021-11-09T08:20:18.554Z [ERROR] raft: failed to appendEntries to: peer="{Voter 172.16.160.103:4002 172.16.160.103:4002}" error="dial tcp 172.16.160.103:4002: connect: connection refused"
2021-11-09T08:20:18.647Z [ERROR] raft: failed to heartbeat to: peer=172.16.160.103:4002 error="dial tcp 172.16.160.103:4002: connect: connection refused"
2021-11-09T08:20:18.665Z [ERROR] raft: failed to appendEntries to: peer="{Voter 172.16.160.103:4002 172.16.160.103:4002}" error="dial tcp 172.16.160.103:4002: connect: connection refused"
2021-11-09T08:20:18.742Z [ERROR] raft: failed to appendEntries to: peer="{Voter 172.16.160.103:4002 172.16.160.103:4002}" error="dial tcp 172.16.160.103:4002: connect: connection refused"
2021-11-09T08:20:18.853Z [ERROR] raft: failed to heartbeat to: peer=172.16.160.103:4002 error="dial tcp 172.16.160.103:4002: connect: connection refused"
2021-11-09T08:20:18.865Z [ERROR] raft: failed to appendEntries to: peer="{Voter 172.16.160.103:4002 172.16.160.103:4002}" error="dial tcp 172.16.160.103:4002: connect: connection refused"
2021-11-09T08:20:19.023Z [ERROR] raft: failed to heartbeat to: peer=172.16.160.103:4002 error="dial tcp 172.16.160.103:4002: connect: connection refused"
2021-11-09T08:20:19.042Z [ERROR] raft: failed to appendEntries to: peer="{Voter 172.16.160.103:4002 172.16.160.103:4002}" error="dial tcp 172.16.160.103:4002: connect: connection refused"
2021-11-09T08:20:19.043Z [WARN]  raft: failed to contact: server-id=172.16.160.103:4002 time=505.318744ms
```

**Comment by wujushan** [comment]
Hi @otoolep, I find that the problem is that all the nodes cannot join cluster, in discovery mode, after all of them restarted manually.

The command line of launching the rqlite:

> /bin/rqlited -http-addr 0.0.0.0:4001 -raft-addr 0.0.0.0:4002 -http-adv-addr 172.16.160.49:4001 -raft-adv-addr 172.16.160.49:4002 -disco-id e2d437c0-3d66-11ec-8442-8fa65a76e101 -disco-url http://172.16.160.49:10808 -join-attempts 10 -join-interval 5s -on-disk /rqlite/file/data


 I find, in the log, the node can receive the request from the other node.

**received request from node with ID 172.16.160.43:4002, at 172.16.160.43:4002, to join this node**

```
[rqlited] 2021/11/10 03:45:03 rqlited starting, version 6, commit unknown, branch unknown, compiler gc
[rqlited] 2021/11/10 03:45:03 go1.17.2, target architecture is arm, operating system target is linux
[rqlited] 2021/11/10 03:45:03 launch command: /bin/rqlited -http-addr 0.0.0.0:4001 -raft-addr 0.0.0.0:4002 -http-adv-addr 172.16.160.49:4001 -raft-adv-addr 172.16.160.49:4002 -disco-id e2d437c0-3d66-11ec-8442-8fa65a76e101 -disco-url http://172.16.160.49:10808 -join-attempts 10 -join-interval 5s -on-disk /rqlite/file/data
[rqlited] 2021/11/10 03:45:03 Raft TCP mux Listener registered with 1
[rqlited] 2021/11/10 03:45:03 no preexisting node state detected in /rqlite/file/data, node may be bootstrapping
[rqlited] 2021/11/10 03:45:03 registering with Discovery Service at http://172.16.160.49:10808 with ID e2d437c0-3d66-11ec-8442-8fa65a76e101
[mux] 2021/11/10 03:45:03 mux serving on [::]:4002, advertising 172.16.160.49:4002
[discovery] 2021/11/10 03:45:03 discovery client attempting registration of 172.16.160.49:4001 at http://172.16.160.49:10808/e2d437c0-3d66-11ec-8442-8fa65a76e101
[discovery] 2021/11/10 03:45:03 discovery client successfully registered 172.16.160.49:4001 at http://172.16.160.49:10808/e2d437c0-3d66-11ec-8442-8fa65a76e101
[rqlited] 2021/11/10 03:45:03 Discovery Service responded with nodes: [172.16.160.43:4001 172.16.160.49:4001 172.16.160.103:4001]
[rqlited] 2021/11/10 03:45:03 join addresses specified, node is not bootstrapping
[store] 2021/11/10 03:45:03 opening store with node ID 172.16.160.49:4002
[store] 2021/11/10 03:45:03 configured for an on-disk database at /rqlite/file/data/db.sqlite
[store] 2021/11/10 03:45:03 ensuring directory at /rqlite/file/data exists
[store] 2021/11/10 03:45:03 ensuring directory at /rqlite/file/data exists
[store] 2021/11/10 03:45:03 0 preexisting snapshots present
[store] 2021/11/10 03:45:03 first log index: 0, last log index: 0, last command log index: 0:
[store] 2021/11/10 03:45:03 created on-disk database at open
2021-11-10T03:45:03.640Z [INFO]  raft: initial configuration: index=0 servers=[]
[store] 2021/11/10 03:45:03 no cluster bootstrap requested
[cluster] 2021/11/10 03:45:03 service listening on 172.16.160.49:4002
[rqlited] 2021/11/10 03:45:03 Cluster TCP mux Listener registered with 2
2021-11-10T03:45:03.640Z [INFO]  raft: entering follower state: follower="Node at 172.16.160.49:4002 [Follower]" leader=
[http] 2021/11/10 03:45:03 service listening on [::]:4001
[rqlited] 2021/11/10 03:45:03 join addresses are: [172.16.160.43:4001 172.16.160.103:4001]
[cluster-join] 2021/11/10 03:45:03 failed to join cluster at [172.16.160.43:4001 172.16.160.103:4001]: Post "http://172.16.160.103:4001/join": dial tcp 172.16.160.103:4001: connect: connection refused, sleeping 5s before retry
2021-11-10T03:45:05.532Z [WARN]  raft: no known peers, aborting election
[cluster-join] 2021/11/10 03:45:08 failed to join cluster at [172.16.160.43:4001 172.16.160.103:4001]: Post "http://172.16.160.103:4001/join": dial tcp 172.16.160.103:4001: connect: connection refused, sleeping 5s before retry
[store] 2021/11/10 03:45:11 received request from node with ID 172.16.160.43:4002, at 172.16.160.43:4002, to join this node
[cluster-join] 2021/11/10 03:45:13 failed to join cluster at [172.16.160.43:4001 172.16.160.103:4001]: Post "http://172.16.160.103:4001/join": dial tcp 172.16.160.103:4001: connect: connection refused, sleeping 5s before retry
[store] 2021/11/10 03:45:16 received request from node with ID 172.16.160.43:4002, at 172.16.160.43:4002, to join this node
[cluster-join] 2021/11/10 03:45:18 failed to join cluster at [172.16.160.43:4001 172.16.160.103:4001]: Post "http://172.16.160.103:4001/join": dial tcp 172.16.160.103:4001: connect: connection refused, sleeping 5s before retry
[store] 2021/11/10 03:45:21 received request from node with ID 172.16.160.43:4002, at 172.16.160.43:4002, to join this node
[cluster-join] 2021/11/10 03:45:23 failed to join cluster at [172.16.160.43:4001 172.16.160.103:4001]: Post "http://172.16.160.103:4001/join": dial tcp 172.16.160.103:4001: connect: connection refused, sleeping 5s before retry
[store] 2021/11/10 03:45:26 received request from node with ID 172.16.160.43:4002, at 172.16.160.43:4002, to join this node

```

**Comment by otoolep** [comment]
Thanks for the extra information.

To debug this I'm going to new more precise instructions. I'm going to need a step-by-step process for reproducing the issue you are seeing since it's not fully clear to me.

- Show me how you start the cluster at the very start (show me each command)
- What exactly you do next (stop a node, restart all nodes, what)
- What exactly you do next. Do you start all the nodes, just one of them etc.

At each stage show me each command you are running, in the sequence you are running them.

**Comment by wujushan** [comment]
Hi @otoolep,

I have three nodes whose addresses are 172.16.160.43, 192.16.160.49, 172.16.160.103. And rqlite will run in docker container.

First of all, I apply for an disco-id from http://discovery.rqlite.com and then lanuch rqlite with the following command line:

> docker run -ti --name rqlite --rm --network host -e HTTP_ADV_ADDR_IP=172.16.160.49 -e RAFT_ADV_ADDR_IP=172.16.160.49 rqlite/rqlite:test -disco-id d7e32700-429d-11ec-95d3-8937f6952003 -join-attempts 10 -join-interval 5s -on-disk

**After the startup is complete, all nodes can work normally.**

Then, I stop all rqlites and then only start **one of them**,  I can get the log below:

```
[rqlited] 2021/11/11 03:39:26 rqlited starting, version 6, commit unknown, branch unknown, compiler gc
[rqlited] 2021/11/11 03:39:26 go1.17.2, target architecture is arm, operating system target is linux
[rqlited] 2021/11/11 03:39:26 launch command: /bin/rqlited -http-addr 0.0.0.0:4001 -raft-addr 0.0.0.0:4002 -http-adv-addr 172.16.160.49:4001 -raft-adv-addr 172.16.160.49:4002 -disco-id d7e32700-429d-11ec-95d3-8937f6952003 -join-attempts 10 -join-interval 5s -on-disk /rqlite/file/data
[rqlited] 2021/11/11 03:39:26 Raft TCP mux Listener registered with 1
[rqlited] 2021/11/11 03:39:26 no preexisting node state detected in /rqlite/file/data, node may be bootstrapping
[rqlited] 2021/11/11 03:39:26 registering with Discovery Service at http://discovery.rqlite.com with ID d7e32700-429d-11ec-95d3-8937f6952003
[mux] 2021/11/11 03:39:26 mux serving on [::]:4002, advertising 172.16.160.49:4002
[discovery] 2021/11/11 03:39:26 discovery client attempting registration of 172.16.160.49:4001 at http://discovery.rqlite.com/d7e32700-429d-11ec-95d3-8937f6952003
[discovery] 2021/11/11 03:39:26 discovery client redirecting to https://k05nh959gb.execute-api.us-west-2.amazonaws.com/prod/rqliteDiscoHandler/d7e32700-429d-11ec-95d3-8937f6952003
[discovery] 2021/11/11 03:39:26 discovery client attempting registration of 172.16.160.49:4001 at https://k05nh959gb.execute-api.us-west-2.amazonaws.com/prod/rqliteDiscoHandler/d7e32700-429d-11ec-95d3-8937f6952003
[discovery] 2021/11/11 03:39:27 discovery client successfully registered 172.16.160.49:4001 at https://k05nh959gb.execute-api.us-west-2.amazonaws.com/prod/rqliteDiscoHandler/d7e32700-429d-11ec-95d3-8937f6952003
[rqlited] 2021/11/11 03:39:27 Discovery Service responded with nodes: [172.16.160.103:4001 172.16.160.43:4001 172.16.160.49:4001]
[rqlited] 2021/11/11 03:39:27 join addresses specified, node is not bootstrapping
[store] 2021/11/11 03:39:27 opening store with node ID 172.16.160.49:4002
[store] 2021/11/11 03:39:27 configured for an on-disk database at /rqlite/file/data/db.sqlite
[store] 2021/11/11 03:39:27 ensuring directory at /rqlite/file/data exists
[store] 2021/11/11 03:39:27 ensuring directory at /rqlite/file/data exists
[store] 2021/11/11 03:39:27 0 preexisting snapshots present
[store] 2021/11/11 03:39:27 first log index: 0, last log index: 0, last command log index: 0:
[store] 2021/11/11 03:39:27 created on-disk database at open
2021-11-11T03:39:27.650Z [INFO]  raft: initial configuration: index=0 servers=[]
[store] 2021/11/11 03:39:27 no cluster bootstrap requested
[cluster] 2021/11/11 03:39:27 service listening on 172.16.160.49:4002
2021-11-11T03:39:27.650Z [INFO]  raft: entering follower state: follower="Node at 172.16.160.49:4002 [Follower]" leader=
[rqlited] 2021/11/11 03:39:27 Cluster TCP mux Listener registered with 2
[http] 2021/11/11 03:39:27 service listening on [::]:4001
[rqlited] 2021/11/11 03:39:27 join addresses are: [172.16.160.103:4001 172.16.160.43:4001]
[cluster-join] 2021/11/11 03:39:27 failed to join cluster at [172.16.160.103:4001 172.16.160.43:4001]: Post "http://172.16.160.43:4001/join": dial tcp 172.16.160.43:4001: connect: connection refused, sleeping 5s before retry
2021-11-11T03:39:28.993Z [WARN]  raft: no known peers, aborting election
[cluster-join] 2021/11/11 03:39:32 failed to join cluster at [172.16.160.103:4001 172.16.160.43:4001]: Post "http://172.16.160.43:4001/join": dial tcp 172.16.160.43:4001: connect: connection refused, sleeping 5s before retry
[cluster-join] 2021/11/11 03:39:37 failed to join cluster at [172.16.160.103:4001 172.16.160.43:4001]: Post "http://172.16.160.43:4001/join": dial tcp 172.16.160.43:4001: connect: connection refused, sleeping 5s before retry
[cluster-join] 2021/11/11 03:39:42 failed to join cluster at [172.16.160.103:4001 172.16.160.43:4001]: Post "http://172.16.160.43:4001/join": dial tcp 172.16.160.43:4001: connect: connection refused, sleeping 5s before retry

```

Finally, I start all the rqlites at the same time, I can get the log below:
 
```
[rqlited] 2021/11/11 03:40:47 rqlited starting, version 6, commit unknown, branch unknown, compiler gc
[rqlited] 2021/11/11 03:40:47 go1.17.2, target architecture is arm, operating system target is linux
[rqlited] 2021/11/11 03:40:47 launch command: /bin/rqlited -http-addr 0.0.0.0:4001 -raft-addr 0.0.0.0:4002 -http-adv-addr 172.16.160.49:4001 -raft-adv-addr 172.16.160.49:4002 -disco-id d7e32700-429d-11ec-95d3-8937f6952003 -join-attempts 10 -join-interval 5s -on-disk /rqlite/file/data
[rqlited] 2021/11/11 03:40:47 Raft TCP mux Listener registered with 1
[rqlited] 2021/11/11 03:40:47 no preexisting node state detected in /rqlite/file/data, node may be bootstrapping
[rqlited] 2021/11/11 03:40:47 registering with Discovery Service at http://discovery.rqlite.com with ID d7e32700-429d-11ec-95d3-8937f6952003
[discovery] 2021/11/11 03:40:47 discovery client attempting registration of 172.16.160.49:4001 at http://discovery.rqlite.com/d7e32700-429d-11ec-95d3-8937f6952003
[mux] 2021/11/11 03:40:47 mux serving on [::]:4002, advertising 172.16.160.49:4002
[discovery] 2021/11/11 03:40:48 discovery client redirecting to https://k05nh959gb.execute-api.us-west-2.amazonaws.com/prod/rqliteDiscoHandler/d7e32700-429d-11ec-95d3-8937f6952003
[discovery] 2021/11/11 03:40:48 discovery client attempting registration of 172.16.160.49:4001 at https://k05nh959gb.execute-api.us-west-2.amazonaws.com/prod/rqliteDiscoHandler/d7e32700-429d-11ec-95d3-8937f6952003
[discovery] 2021/11/11 03:40:48 discovery client successfully registered 172.16.160.49:4001 at https://k05nh959gb.execute-api.us-west-2.amazonaws.com/prod/rqliteDiscoHandler/d7e32700-429d-11ec-95d3-8937f6952003
[rqlited] 2021/11/11 03:40:48 Discovery Service responded with nodes: [172.16.160.103:4001 172.16.160.43:4001 172.16.160.49:4001]
[rqlited] 2021/11/11 03:40:48 join addresses specified, node is not bootstrapping
[store] 2021/11/11 03:40:48 opening store with node ID 172.16.160.49:4002
[store] 2021/11/11 03:40:48 configured for an on-disk database at /rqlite/file/data/db.sqlite
[store] 2021/11/11 03:40:48 ensuring directory at /rqlite/file/data exists
[store] 2021/11/11 03:40:48 ensuring directory at /rqlite/file/data exists
[store] 2021/11/11 03:40:48 0 preexisting snapshots present
[store] 2021/11/11 03:40:48 first log index: 0, last log index: 0, last command log index: 0:
[store] 2021/11/11 03:40:48 created on-disk database at open
2021-11-11T03:40:48.960Z [INFO]  raft: initial configuration: index=0 servers=[]
[store] 2021/11/11 03:40:48 no cluster bootstrap requested
2021-11-11T03:40:48.960Z [INFO]  raft: entering follower state: follower="Node at 172.16.160.49:4002 [Follower]" leader=
[cluster] 2021/11/11 03:40:48 service listening on 172.16.160.49:4002
[rqlited] 2021/11/11 03:40:48 Cluster TCP mux Listener registered with 2
[http] 2021/11/11 03:40:48 service listening on [::]:4001
[rqlited] 2021/11/11 03:40:48 join addresses are: [172.16.160.103:4001 172.16.160.43:4001]
[cluster-join] 2021/11/11 03:40:48 failed to join cluster at [172.16.160.103:4001 172.16.160.43:4001]: Post "http://172.16.160.43:4001/join": dial tcp 172.16.160.43:4001: connect: connection refused, sleeping 5s before retry
[store] 2021/11/11 03:40:49 received request from node with ID 172.16.160.43:4002, at 172.16.160.43:4002, to join this node
2021-11-11T03:40:50.747Z [WARN]  raft: no known peers, aborting election
[cluster-join] 2021/11/11 03:40:53 failed to join cluster at [172.16.160.103:4001 172.16.160.43:4001]: failed to join, node returned: 503 Service Unavailable: (leader not found
), sleeping 5s before retry
[store] 2021/11/11 03:40:54 received request from node with ID 172.16.160.43:4002, at 172.16.160.43:4002, to join this node
[cluster-join] 2021/11/11 03:40:58 failed to join cluster at [172.16.160.103:4001 172.16.160.43:4001]: failed to join, node returned: 503 Service Unavailable: (leader not found
), sleeping 5s before retry
[store] 2021/11/11 03:40:59 received request from node with ID 172.16.160.43:4002, at 172.16.160.43:4002, to join this node
[store] 2021/11/11 03:41:02 received request from node with ID 172.16.160.103:4002, at 172.16.160.103:4002, to join this node
[cluster-join] 2021/11/11 03:41:03 failed to join cluster at [172.16.160.103:4001 172.16.160.43:4001]: failed to join, node returned: 503 Service Unavailable: (leader not found
), sleeping 5s before retry
[store] 2021/11/11 03:41:04 received request from node with ID 172.16.160.43:4002, at 172.16.160.43:4002, to join this node
[store] 2021/11/11 03:41:07 received request from node with ID 172.16.160.103:4002, at 172.16.160.103:4002, to join this node
[cluster-join] 2021/11/11 03:41:08 failed to join cluster at [172.16.160.103:4001 172.16.160.43:4001]: failed to join, node returned: 503 Service Unavailable: (leader not found
), sleeping 5s before retry
[store] 2021/11/11 03:41:09 received request from node with ID 172.16.160.43:4002, at 172.16.160.43:4002, to join this node
[store] 2021/11/11 03:41:12 received request from node with ID 172.16.160.103:4002, at 172.16.160.103:4002, to join this node
[cluster-join] 2021/11/11 03:41:13 failed to join cluster at [172.16.160.103:4001 172.16.160.43:4001]: failed to join, node returned: 503 Service Unavailable: (leader not found
), sleeping 5s before retry
[store] 2021/11/11 03:41:14 received request from node with ID 172.16.160.43:4002, at 172.16.160.43:4002, to join this node
[store] 2021/11/11 03:41:17 received request from node with ID 172.16.160.103:4002, at 172.16.160.103:4002, to join this node
[cluster-join] 2021/11/11 03:41:19 failed to join cluster at [172.16.160.103:4001 172.16.160.43:4001]: failed to join, node returned: 503 Service Unavailable: (leader not found
), sleeping 5s before retry
[store] 2021/11/11 03:41:19 received request from node with ID 172.16.160.43:4002, at 172.16.160.43:4002, to join this node
[store] 2021/11/11 03:41:23 received request from node with ID 172.16.160.103:4002, at 172.16.160.103:4002, to join this node
[cluster-join] 2021/11/11 03:41:24 failed to join cluster at [172.16.160.103:4001 172.16.160.43:4001]: failed to join, node returned: 503 Service Unavailable: (leader not found
), sleeping 5s before retry
[store] 2021/11/11 03:41:24 received request from node with ID 172.16.160.43:4002, at 172.16.160.43:4002, to join this node
[store] 2021/11/11 03:41:28 received request from node with ID 172.16.160.103:4002, at 172.16.160.103:4002, to join this node
[cluster-join] 2021/11/11 03:41:29 failed to join cluster at [172.16.160.103:4001 172.16.160.43:4001]: failed to join, node returned: 503 Service Unavailable: (leader not found
), sleeping 5s before retry
[store] 2021/11/11 03:41:29 received request from node with ID 172.16.160.43:4002, at 172.16.160.43:4002, to join this node
[store] 2021/11/11 03:41:33 received request from node with ID 172.16.160.103:4002, at 172.16.160.103:4002, to join this node

```

**Comment by otoolep** [comment]
Thanks, this detail is helpful.

Are you using persistent disks for the containers? You mount `/rqlite/file` to a place with storage that lasts part restart? And the IP addresses remain the same after the restart?

**Comment by otoolep** [comment]
I notice you are passing `--rm` to your Docker containers, which means they are removed after stopping. This means no state is kept between restarts. The behaviour of rqlite -- removing state completely between cluster restarts but re-using the same Discovery ID -- is undefined. In other words I don't know what will happen -- when the Discovery service was designed, it wasn't for such a case. 

I might be able to make it work, but I will need to look into it.

Can you tell me what you're trying to do? Are you testing cluster recovery after a restart? Or are you trying to make sure a new cluster can find itself everytime? If the latter, you may need a new Disco ID each time, since after a restart *it's a completely new cluster* (assuming you're not using persistent disks -- are you?).

**Comment by wujushan** [comment]
Hi @otoolep，
Thanks for your reply. 

After I use persistent disk, the cluster works normally, but it needs at least two nodes to start up. **If I restart only one node after stopping all nodes, this node will stop itself after all attempts have been made**. **And in addition, I find if the node try to join a new cluster with a new Discovery ID, it will ignore the information given by discovery service but use the old state data. It has to remove the state in first before joining a new cluster.** Are these normal behaviors for rqlite?

The log below is printed by the node whose wants to join a  new cluster.

```
[rqlited] 2021/11/12 02:17:18 rqlited starting, version 6, commit unknown, branch unknown, compiler gc
[rqlited] 2021/11/12 02:17:18 go1.17.2, target architecture is amd64, operating system target is linux
[rqlited] 2021/11/12 02:17:18 launch command: /bin/rqlited -http-addr 0.0.0.0:4001 -raft-addr 0.0.0.0:4002 -http-adv-addr 172.16.160.103:4001 -raft-adv-addr 172.16.160.103:4002 -disco-id 8197ee74-435e-11ec-be03-a701ec177f76 -join-attempts 10 -join-interval 5s -raft-apply-timeout 120s /rqlite/file/data
[rqlited] 2021/11/12 02:17:18 Raft TCP mux Listener registered with 1
[rqlited] 2021/11/12 02:17:18 preexisting node state detected in /rqlite/file/data
[rqlited] 2021/11/12 02:17:18 registering with Discovery Service at http://discovery.rqlite.com with ID 8197ee74-435e-11ec-be03-a701ec177f76
[discovery] 2021/11/12 02:17:18 discovery client attempting registration of 172.16.160.103:4001 at http://discovery.rqlite.com/8197ee74-435e-11ec-be03-a701ec177f76
[mux] 2021/11/12 02:17:18 mux serving on [::]:4002, advertising 172.16.160.103:4002
[discovery] 2021/11/12 02:17:18 discovery client redirecting to https://k05nh959gb.execute-api.us-west-2.amazonaws.com/prod/rqliteDiscoHandler/8197ee74-435e-11ec-be03-a701ec177f76
[discovery] 2021/11/12 02:17:18 discovery client attempting registration of 172.16.160.103:4001 at https://k05nh959gb.execute-api.us-west-2.amazonaws.com/prod/rqliteDiscoHandler/8197ee74-435e-11ec-be03-a701ec177f76
[discovery] 2021/11/12 02:17:19 discovery client successfully registered 172.16.160.103:4001 at https://k05nh959gb.execute-api.us-west-2.amazonaws.com/prod/rqliteDiscoHandler/8197ee74-435e-11ec-be03-a701ec177f76
[rqlited] 2021/11/12 02:17:19 Discovery Service responded with nodes: [172.16.160.103:4001]
[rqlited] 2021/11/12 02:17:19 no join addresses set
[store] 2021/11/12 02:17:19 opening store with node ID 172.16.160.103:4002
[store] 2021/11/12 02:17:19 configured for an in-memory database
[store] 2021/11/12 02:17:19 ensuring directory for Raft exists at /rqlite/file/data
[store] 2021/11/12 02:17:19 0 preexisting snapshots present
[store] 2021/11/12 02:17:19 first log index: 1, last log index: 7, last command log index: 0:
[store] 2021/11/12 02:17:19 created in-memory database at open
2021-11-12T02:17:19.560Z [INFO]  raft: initial configuration: index=4 servers="[{Suffrage:Voter ID:172.16.160.103:4002 Address:172.16.160.103:4002} {Suffrage:Voter ID:172.16.160.49:4002 Address:172.16.160.49:4002} {Suffrage:Voter ID:172.16.160.43:4002 Address:172.16.160.43:4002}]"
[store] 2021/11/12 02:17:19 no cluster bootstrap requested
[cluster] 2021/11/12 02:17:19 service listening on 172.16.160.103:4002
[rqlited] 2021/11/12 02:17:19 Cluster TCP mux Listener registered with 2
[http] 2021/11/12 02:17:19 service listening on [::]:4001
2021-11-12T02:17:19.561Z [INFO]  raft: entering follower state: follower="Node at 172.16.160.103:4002 [Follower]" leader=
2021-11-12T02:17:20.902Z [WARN]  raft: heartbeat timeout reached, starting election: last-leader=
2021-11-12T02:17:20.902Z [INFO]  raft: entering candidate state: node="Node at 172.16.160.103:4002 [Candidate]" term=90
2021-11-12T02:17:20.985Z [ERROR] raft: failed to make requestVote RPC: target="{Voter 172.16.160.49:4002 172.16.160.49:4002}" error="dial tcp 172.16.160.49:4002: connect: connection refused"
2021-11-12T02:17:20.985Z [ERROR] raft: failed to make requestVote RPC: target="{Voter 172.16.160.43:4002 172.16.160.43:4002}" error="dial tcp 172.16.160.43:4002: connect: connection refused"
2021-11-12T02:17:22.154Z [WARN]  raft: Election timeout reached, restarting election
2021-11-12T02:17:22.154Z [INFO]  raft: entering candidate state: node="Node at 172.16.160.103:4002 [Candidate]" term=91
2021-11-12T02:17:22.218Z [ERROR] raft: failed to make requestVote RPC: target="{Voter 172.16.160.49:4002 172.16.160.49:4002}" error="dial tcp 172.16.160.49:4002: connect: connection refused"
2021-11-12T02:17:22.218Z [ERROR] raft: failed to make requestVote RPC: target="{Voter 172.16.160.43:4002 172.16.160.43:4002}" error="dial tcp 172.16.160.43:4002: connect: connection refused"
2021-11-12T02:17:23.915Z [WARN]  raft: Election timeout reached, restarting election
2021-11-12T02:17:23.916Z [INFO]  raft: entering candidate state: node="Node at 172.16.160.103:4002 [Candidate]" term=92
2021-11-12T02:17:23.977Z [ERROR] raft: failed to make requestVote RPC: target="{Voter 172.16.160.43:4002 172.16.160.43:4002}" error="dial tcp 172.16.160.43:4002: connect: connection refused"
2021-11-12T02:17:23.977Z [ERROR] raft: failed to make requestVote RPC: target="{Voter 172.16.160.49:4002 172.16.160.49:4002}" error="dial tcp 172.16.160.49:4002: connect: connection refused"

```

**Comment by otoolep** [comment]
> If I restart only one node after stopping all nodes, this node will stop itself after all attempts have been made

Yes, this is expected. If you've created a 3-node cluster, stop it all nodes in the cluster, it will not come up unless a quorum of nodes are available -- in this case 2 nodes. This is the way Raft works.

>And in addition, I find if the node try to join a new cluster with a new Discovery ID, it will ignore the information given by discovery service but use the old state data

Yes, again this is by design. Discovery is effectively ignored once nodes join a cluster. From then on the nodes use information about the cluster they have gained from joining the cluster, not the what is in the Discovery service.

If you want to restart a cluster but only with 1 node, you must tell the node not to expect the other nodes to appear. However to do this you must have a leader, but a leader cannot be elected after restart unless you have a quorum (see above). But if you wish to override and go back to one node, you must follow this process:

https://github.com/rqlite/rqlite/blob/master/DOC/CLUSTER_MGMT.md#recovering-a-cluster-that-has-permanently-lost-quorum

Basically you need to create a file under the single node you wish to start, telling it that its the only node.

**Comment by otoolep** [comment]
See also https://github.com/rqlite/rqlite/blob/master/DOC/DISCOVERY.md#caveats

**Comment by otoolep** [comment]
I think this answers your questions. Please close this issue if so -- thanks.

**Comment by otoolep** [comment]
It's a minor bug that it's not clearer that the Discovery service does not appear to be ignored: https://github.com/rqlite/rqlite/issues/347

**Comment by wujushan** [comment]
@otoolep

Thanks a lot.
